libkbfs/block_journal.go

```diff
@@ -162,65 +185,73 @@ func (j *bserverTlfJournal) readJournalLocked() (
 	}
 
 	for i := first; i <= last; i++ {
-		e, err := j.readJournalEntryLocked(i)
+		e, err := j.readJournalEntry(i)
 		if err != nil {
 			return nil, err
 		}
 
-		blockRefs := refs[e.ID]
-		if blockRefs == nil {
-			blockRefs = make(blockRefMap)
-			refs[e.ID] = blockRefs
-		}
-
+		// Handle single ops separately.
 		switch e.Op {
 		case blockPutOp, addRefOp:
-			if len(e.Contexts) != 1 {
-				return nil, fmt.Errorf(
-					"Op %s for id=%s doesn't have exactly one context: %v",
-					e.Op, e.ID, e.Contexts)
+			id, context, err := e.getSingleContext()
+			if err != nil {
+				return nil, err
 			}
 
-			blockRefs.put(e.Contexts[0], liveBlockRef)
-
-		case removeRefsOp:
-			for _, context := range e.Contexts {
-				delete(blockRefs, context.GetRefNonce())
+			blockRefs := refs[id]
+			if blockRefs == nil {
+				blockRefs = make(blockRefMap)
+				refs[id] = blockRefs
 			}
 
-		case archiveRefsOp:
-			for _, context := range e.Contexts {
-				blockRefs.put(context, archivedBlockRef)
+			blockRefs.put(context, liveBlockRef)
+			continue
+		}
+
+		for id, idContexts := range e.Contexts {
+			blockRefs := refs[id]
+			if blockRefs == nil {
+				blockRefs = make(blockRefMap)
+				refs[id] = blockRefs
 			}
 
-		default:
-			return nil, fmt.Errorf("Unknown op %s", e.Op)
+			switch e.Op {
+			case removeRefsOp:
+				for _, context := range idContexts {
+					delete(blockRefs, context.GetRefNonce())
```

Comment:
Any reason not to use `blockRefs.remove()`? It does some extra error checking that I _think_ should still be applicable to the journal.